### PR TITLE
Add Language authoring 2022-05-01 APIs

### PR DESCRIPTION
**Pull request by moabba**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [X] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by moabba** [comment]
This PR is still pending the examples.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:31]: Is the /authoring segment necessary? Looking at questionanswering-authoring.json from 2021-05-01-preview, it establishes a pattern where you'd just use /analyze-conversations/projects instead. This won't conflict with your inference api /:analyze-conversations since it's an operation endpoint.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:218]: We should already have a parameter for this in common.json. Or is that in-body? If it is in-body, you could add one for the query (though we use "Parameter" as a suffix for consistency).

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:232]: Should have descriptions for GAs at least.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:274]: Nit: typically just call this "body".

Comment (heaths) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:274]: Some instances below as well.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:745]: Nit: could define a shared parameter for import/export.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:918]: Nit: could also define a shared parameter for this as well. Used in the endpoint above.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:962]:
```suggestion
        "description": "Gets the evaluation summary of a trained model. The summary includes high level performance measurements of the model e.g., F1, Precision, Recall, etc.",
```

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1135]: Why does this follow a different pattern? Wouldn't ../delete/jobs/... be more consistent with other job status endpoints?

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1177]: Are these BCP-47 languages? IF so, would be good to specify that as we've been doing in other APIs. If not, still specify what, please.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1235]: Same here. Wherever you mention the language, it's good to specify in what form it should be e.g., BCP-47.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1338]: Nit: most of the enum values I'm seeing in other Language service swaggers are using PascalCase. Would be good to be consistent throughout Language.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1344]: The enum values should be described. See https://github.com/Azure/autorest/tree/main/docs/extensions#x-ms-enum

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1333]: I'm not seeing descriptions for a lot of these models. Not only does this present a less-than-ideal UX for REST API docs, but for SDK docs as well - none of the generated classes (pretty much all of them) will be documented.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1363]: I'm unclear how this relates to `rawValue`, if at all. Is this just `rawValue` * 100? If not, throughout Language we've been representing percentages as 0.0 to 1.0 and you should specify the range. This seems more common in swaggers within and without Microsoft.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1470]: Why isn't this an enum then? Using PascalCase for acceptable values here also seems to conflict with most/all enums I'm seeing elsewhere in this swagger. Based on other Language swaggers, I recommend using PascalCase.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1524]: Are there any differences across job states and status? Couldn't they be folded into a single definition. IIRC, for 2022-05-15-preview someone moved all the job state models and related properties to common.json for unification.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:3013]: Nit: we've been using "Parameter" as a suffix.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:3028]: Could you use the `StringIndexType` in common.json instead?

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json:4]: Same comments here as for conversationalysisauthoirng.json.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json:31]: Same comment as for the other swagger about /authoring and file name, though I see from https://github.com/Azure/azure-rest-api-specs/blob/4caf2df6566101e2a17a85b7afd856aa534c1429/specification/cognitiveservices/data-plane/Language/preview/2022-03-01-preview/textanalytics.json#L31 this file should probably be called _textanalytics-authoring.json_.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json:625]: As noted in the other swagger, I suspect all these JobStates are the same. May be better to refactor to common.json. Any specifics could use an `allOf` reference to that and add properties as needed.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json
Comment (heaths) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json:837]: Like the other swagger, I imagine this gets reused a lot. At least a local parameter might be nice to make reviews easier.

**Review by heaths — CHANGES_REQUESTED** [review]
Comments are largely the same across both swaggers. Overall, though, I'm concerned we haven't previewed these yet and are planning to go GA already.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (johanste) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1489]: This is really the same as `ConversationalAnalysisAuthoringProjectMetadata` and should be "folded" into a single model with `readOnly` properties. Unless I'm missing something?

Comment (johanste) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1489]: Never mind, I was missing something. There is a difference in requiredness between the create and read models - which, in turn, is due to the use of PATCH for upsert.

**Comment by moabba** [comment]
> Comments are largely the same across both swaggers. Overall, though, I'm concerned we haven't previewed these yet and are planning to go GA already.

We had 2021-11-01-preview, 2022-02-01-preview and 2022-03-01-preview APIs published for the authoring. The difference between all of these API versions were made according to the feedback we get from our reviews with the API board. This final GA swagger file is highly similar to our 2022-03-01-preview APIs, except for a few changes that were also requested by the API board.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:31]: Initially we didn't have the "/authoring/" prefix, but this caused conflicts in our application gateway (APIM) where the different language configurations would conflict with each other.

E.g. The authoring would use /language/analyze-text/projects as a prefix. The inference would use /language as a prefix and have an operation with path analyze-text/jobs. This operation would conflict with the authoring's prefix and return 404. We have a thread talking about this and a discussion was done a while back. @johanste was part of the discussion and attended the meeting that reached the "/language/authoring" as a prefix conclusion.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:3028]: The authoring still doesn't support all the StringIndexType that exists in the common.json file. That's why it's defined here with a subset of the values. We're planning to support the same set of values in the future. Then, we'll reference the one in common.json instead.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json:837]: Should we move all the parameters to the parameters section? Or only the re-used ones?

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1489]: Actually, you're right. They are the same, except for the "DateTime" properties (E.g. `lastModifiedDateTime`). In the PATCH body, the properties are marked "required" assuming that it's a Create request (i.e. we're patching an empty object).

Should we make these properties optional, given that it's a merge-patch+json request, or should we mark them as required assuming that the resource didn't exist before this request?

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json
Comment (heaths) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json:837]: I recommend - though not critical - moving commonly used ones. Whether that's 2 or 3+ uses is up to you (or none).

**Comment by heaths** [comment]
> We had 2021-11-01-preview, 2022-02-01-preview and 2022-03-01-preview APIs published for the authoring. The difference between all of these API versions were made according to the feedback we get from our reviews with the API board. This final GA swagger file is highly similar to our 2022-03-01-preview APIs, except for a few changes that were also requested by the API board.

Looking in https://github.com/Azure/azure-rest-api-specs/tree/main/specification/cognitiveservices/data-plane/Language/preview/2022-03-01-preview, for example, I don't see any authoring. I'm also a codeowner on that directory (for review purposes) and don't recall seeing any authoring APIs. Where they actually merged to main (published)?

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1489]: And just to make sure that I understand correctly. We can mark properties as `readonly` to indicate that it's not a user input, and it's a server generated value. This applies to generated `DateTime` and maybe the `modelId` (system generated).

Or asking the question in a different way. Should we use `readonly` for all the responses that we return if it's not updatable? E.g. The evaluation results are responses only, should we make all its properties `readonly`?

**Comment by moabba** [comment]
> > We had 2021-11-01-preview, 2022-02-01-preview and 2022-03-01-preview APIs published for the authoring. The difference between all of these API versions were made according to the feedback we get from our reviews with the API board. This final GA swagger file is highly similar to our 2022-03-01-preview APIs, except for a few changes that were also requested by the API board.
> 
> Looking in https://github.com/Azure/azure-rest-api-specs/tree/main/specification/cognitiveservices/data-plane/Language/preview/2022-03-01-preview, for example, I don't see any authoring. I'm also a codeowner on that directory (for review purposes) and don't recall seeing any authoring APIs. Where they actually merged to main (published)?
> 
> _questionanswering-authoring.json_ was the only authoring APIs I recall viewing or could find in main.

No, they weren't published. If you recall the previous PR that was closed, was for 2022-03-01-preview. It originally was for 2022-02-01-preview, but we moved it to 2022-03-01-preview as an evolution of the previous API, after applying the feedback we received in the PR and from the board.

**Comment by heaths** [comment]
> No, they weren't published. If you recall the previous PR that was closed, was for 2022-03-01-preview. It originally was for 2022-02-01-preview, but we moved it to 2022-03-01-preview as an evolution of the previous API, after applying the feedback we received in the PR and from the board.

It concerns me that these were never actually published, nor have we ever generated SDKs for them and published betas. We would like to ship at least one beta of the SDKs, but any feedback we get that might affect the service may not be possible if these swaggers (and the backing service implementations) GA without previews.

/cc @mayurid

**Comment by moabba** [comment]
> > No, they weren't published. If you recall the previous PR that was closed, was for 2022-03-01-preview. It originally was for 2022-02-01-preview, but we moved it to 2022-03-01-preview as an evolution of the previous API, after applying the feedback we received in the PR and from the board.
> 
> It concerns me that these were never actually published, nor have we ever generated SDKs for them and published betas. We would like to ship at least one beta of the SDKs, but any feedback we get that might affect the service may not be possible if these swaggers (and the backing service implementations) GA without previews.
> 
> /cc @mayurid

@cahann can you please chime in here?

**Comment by cahann** [comment]
@heaths this has never been a requirement, we had exception for authoring SDKs. We updated the APIs in the past iterations based on the customer feedback which is the same process the question answering is following.

File: specification/cognitiveservices/data-plane/Language/readme.md
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/readme.md:33]: The inference APIs will be added in another PR by Chong or someone on his team.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1135]: Because there's no `delete` operation like the other ones.
- For `:train` we have `/train/{jobId}`.
- For `:export` we have `/export/{jobId}`.

You'll find that this is the same approach take in QnA.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1338]: QnA went GA with all the enums being camelCase, except for the error codes and the string index type. Text Analytics has a mix between camelCase and PascalCase. If we want to have the consistency across the services, then TA should convert all their enums to be camelCase instead.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1363]: `rawValue` represents the value that created the percentages. For example, the number of documents that were correctly predicted is 3 out of 4, then `rawValue` would be 3 and `normalizedValue` would be 75.

Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1363]: Since this doesn't represent a score, and we also use percentage to be 0-100 in the JobState to represent the progress, I would prefer to keep the values 0-100 rather than 0-1.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json:1470]: Today we support only these two pre-defined variables. But down the road we'll allow user input as well. So, it'll change from an enum to a user-defined string. That's why it's done this way for now.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json:31]: Replied to the comment above with the reason for the word `authoring` in the path.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/customtextauthoring.json]: Consolidated the JobState. Will move to common.json after agreeing with the other Language services.

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json
Comment (moabba) [reply specification/cognitiveservices/data-plane/Language/stable/2022-05-01/conversationalanalysisauthoring.json]: The authoring supports a subset of the StringIndexType values as mentioned in the other comment.

**Review by heaths — APPROVED** [review]

File: specification/cognitiveservices/data-plane/Language/stable/2022-05-01/analyzetext-authoring.json
Comment (mikekistler) [inline specification/cognitiveservices/data-plane/Language/stable/2022-05-01/analyzetext-authoring.json:1970]: Property names should be lower camel case, e.g. "projectFileVersion". Kebab case is problematic in languages like JavaScript.

**Comment by heaths** [comment]
@lmazuel can you merge this? It's the same issue as it has been - making the `error` property required, which it was always provided in an error response anyway. Hopefully, since this is a GA of common.json, we'll stop seeing that issue after this feature branch is merged to main.

**Pull request merged**
